### PR TITLE
controller/kubefedcluster: fixes for delete handling and suppress changes due to status updates

### PR DESCRIPTION
**Pull request by pjferrell**
controller/kubefedcluster: add handling for deletes with tombstone and prevent constant processing of cluster changes when status is updated (by caching the last observed spec and comparing on informer updates)

Related issue:
https://github.com/kubernetes-sigs/kubefed/issues/950

### CODE REVIEW

File: pkg/controller/kubefedcluster/controller.go [hunk 1]
```diff
@@ -98,43 +102,71 @@ func newClusterController(config *util.ControllerConfig, clusterHealthCheckConfi
 		&fedv1b1.KubeFedCluster{},
 		util.NoResyncPeriod,
 		&cache.ResourceEventHandlerFuncs{
-			DeleteFunc: cc.delFromClusterSet,
-			AddFunc:    cc.addToClusterSet,
+			DeleteFunc: func(obj interface{}) {
+				castObj, ok := obj.(*fedv1b1.KubeFedCluster)
+				if !ok {
```
Comment (marun) [inline pkg/controller/kubefedcluster/controller.go]: I appreciate your attention to detail here, but even this level of rigor won't preclude missing deletion events.  Would you be up for removing these changes from the PR in favor of a separate PR that used a finalizer to ensure reliable cleanup on deletion? That change seems orthogonal to eliminating updates in response to status changes.

File: pkg/controller/kubefedcluster/controller.go [hunk 2]
```diff
@@ -98,43 +102,71 @@ func newClusterController(config *util.ControllerConfig, clusterHealthCheckConfi
...
+				cc.addToClusterSet(castObj)
+			},
 			UpdateFunc: func(oldObj, newObj interface{}) {
-				cc.delFromClusterSet(oldObj)
-				cc.addToClusterSet(newObj)
+				var clusterChanged bool
+				cluster := newObj.(*fedv1b1.KubeFedCluster)
+				cc.mu.Lock()
+				clusterData, ok := cc.clusterDataMap[cluster.Name]
+				if ok && !equality.Semantic.DeepEqual(clusterData.cachedObj.Spec, cluster.Spec) {
```
Comment (marun) [inline pkg/controller/kubefedcluster/controller.go]: Is the use of `metadata.generation` not sufficient to detect when non-status changes occur?

Comment (font) [inline pkg/controller/kubefedcluster/controller.go]: Instead of a cached obj why not compare with `oldObj` passed in?

Comment (marun) [reply pkg/controller/kubefedcluster/controller.go]: I'm assuming it's because there is no guarantee that the oldObj that gets passed in is the one that the controller last reconciled.

**Comment by marun** [comment]
@pjferrell Thank you, this PR represents a nice improvement in the reliability of the cluster controller.

/approve
